File: edit-service/application_package_test.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func approvedPackageFixture(t *testing.T) (localBuildConfig, string) {
	t.Helper()
	root := t.TempDir()
	profile := buildProfile{SiteURL: "https://example.test", Base: "/", Audience: "internal", EditAPI: "https://edit.example.test", CanvasBackend: "true"}
	files := []map[string]any{}
	for _, path := range []string{"private/renderer.mjs", "private/template.html", "private/template-docs.html", "public/.vite/manifest.json", "private/commands/prepare-content.mjs", "private/commands/coordinate.mjs", "private/commands/staged.mjs", "private/commands/application-package.mjs", "private/commands/client-assets.mjs", "private/commands/content-output.mjs", "private/commands/build-item-history.mjs", "private/commands/check-demo.mjs", "private/commands/check-document-links.mjs", "private/commands/check-item-history.mjs", "private/commands/validate_items.py"} {
		if err := writeConfined(root, path, []byte("fixture")); err != nil {
			t.Fatal(err)
		}
		hash := sha256.Sum256([]byte("fixture"))
		files = append(files, map[string]any{"path": path, "hash": hex.EncodeToString(hash[:]), "size": 7})
	}
	source := strings.Repeat("a", 40)
	manifest := map[string]any{"type": "module", "protocol": 1, "source": source, "nodeMajor": 24, "profile": profile, "files": files, "base": "/", "audience": "internal", "dependencyDigest": strings.Repeat("b", 64)}
	bytes, _ := json.Marshal(manifest)
	os.WriteFile(filepath.Join(root, "package.json"), bytes, 0600)
	sum := sha256.Sum256(bytes)
	pointer := filepath.Join(t.TempDir(), "approved.json")
	if err := writeJSONAtomic(pointer, map[string]any{"directory": root, "digest": hex.EncodeToString(sum[:]), "source": source, "repo": "example/roadmap", "workflowRunId": 123}); err != nil {
		t.Fatal(err)
	}
	return localBuildConfig{ApplicationPointer: pointer, Profile: profile}, root
}

func TestApprovedApplicationPackage(t *testing.T) {
	for _, scenario := range []string{"valid", "tampered", "wrong-repo", "wrong-profile", "symlink", "unlisted", "missing"} {
		t.Run(scenario, func(t *testing.T) {
			cfg, root := approvedPackageFixture(t)
			repo := "example/roadmap"
			switch scenario {
			case "tampered":
				os.WriteFile(filepath.Join(root, "private/renderer.mjs"), []byte("changed"), 0600)
			case "wrong-repo":
				repo = "other/roadmap"
			case "wrong-profile":
				cfg.Profile.Audience = "public"
			case "symlink":
				os.Remove(filepath.Join(root, "private/renderer.mjs"))
				os.Symlink("commands/coordinate.mjs", filepath.Join(root, "private/renderer.mjs"))
			case "unlisted":
				os.WriteFile(filepath.Join(root, "private/extra.mjs"), []byte("extra"), 0600)
			case "missing":
				os.Remove(cfg.ApplicationPointer)
			}
			app, err := loadApprovedApplication(cfg, repo)
			if scenario == "valid" {
				if err != nil || app.Source != strings.Repeat("a", 40) {
					t.Fatalf("valid package: %v", err)
				}
			} else if err == nil {
				t.Fatal("untrusted package accepted")
			}
		})
	}
	t.Run("missing-required-demo-check", func(t *testing.T) {
		cfg, root := approvedPackageFixture(t)
		var manifest map[string]any
		if err := readJSON(filepath.Join(root, "package.json"), &manifest); err != nil {
			t.Fatal(err)
		}
		files := manifest["files"].([]any)
		filtered := files[:0]
		for _, value := range files {
			if value.(map[string]any)["path"] != "private/commands/check-demo.mjs" {
				filtered = append(filtered, value)
			}
		}
		manifest["files"] = filtered
		bytes, _ := json.Marshal(manifest)
		if err := os.WriteFile(filepath.Join(root, "package.json"), bytes, 0600); err != nil {
			t.Fatal(err)
		}
		if err := os.Remove(filepath.Join(root, "private/commands/check-demo.mjs")); err != nil {
			t.Fatal(err)
		}
		sum := sha256.Sum256(bytes)
		if err := writeJSONAtomic(cfg.ApplicationPointer, approvedApplication{Directory: root, Digest: hex.EncodeToString(sum[:]), Source: strings.Repeat("a", 40), Repo: "example/roadmap", WorkflowRunID: 123}); err != nil {
			t.Fatal(err)
		}
		if _, err := loadApprovedApplication(cfg, "example/roadmap"); err == nil || !strings.Contains(err.Error(), "incomplete") {
			t.Fatalf("package without demo checker accepted: %v", err)
		}
	})
}

File: site/scripts/application-package.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtemp, mkdir, writeFile, rm, symlink, readFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { createHash } from 'node:crypto';
import { verifyApplicationPackage } from './application-package.mjs';

const hash = bytes => createHash('sha256').update(bytes).digest('hex');
async function fixture(t) {
  const root = await mkdtemp(join(tmpdir(), 'roadmap-package-test-'));
  t.after(() => rm(root, { recursive: true, force: true }));
  const files = [];
  for (const path of ['private/renderer.mjs', 'private/template.html', 'public/.vite/manifest.json']) {
    await mkdir(join(root, path, '..'), { recursive: true });
    await writeFile(join(root, path), 'fixture');
    files.push({ path, hash: hash('fixture'), size: 7 });
  }
  const manifest = { type: 'module', protocol: 1, source: 'a'.repeat(40), base: '/', audience: 'internal',
    nodeMajor: Number(process.versions.node.split('.')[0]), dependencyDigest: 'b'.repeat(64),
    profile: { siteUrl: 'https://example.test', base: '/', audience: 'internal', editApi: '', canvasBackend: '' }, files };
  const save = async () => { const bytes = JSON.stringify(manifest); await writeFile(join(root, 'package.json'), bytes); return hash(bytes); };
  return { root, manifest, save, digest: await save() };
}

test('requires independently supplied digest before trusting executable files', async t => {
  const f = await fixture(t);
  assert.equal((await verifyApplicationPackage(f.root, { digest: f.digest })).source, f.manifest.source);
  await assert.rejects(verifyApplicationPackage(f.root, {}), /digest/i);
  await assert.rejects(verifyApplicationPackage(f.root, { digest: 'c'.repeat(64) }), /digest/i);
  f.manifest.source = 'd'.repeat(40);
  await f.save();
  await assert.rejects(verifyApplicationPackage(f.root, { digest: f.digest }), /digest/i);
});

test('rejects tampering, unlisted files and symlinks', async t => {
  const f = await fixture(t);
  await writeFile(join(f.root, 'private/renderer.mjs'), 'tampered');
  await assert.rejects(verifyApplicationPackage(f.root, { digest: f.digest }), /checksum/i);
  await writeFile(join(f.root, 'private/renderer.mjs'), 'fixture');
  await writeFile(join(f.root, 'private/unlisted.mjs'), 'unlisted');
  await assert.rejects(verifyApplicationPackage(f.root, { digest: f.digest }), /inventory/i);
  await rm(join(f.root, 'private/unlisted.mjs'));
  await rm(join(f.root, 'private/template.html'));
  await symlink('renderer.mjs', join(f.root, 'private/template.html'));
  await assert.rejects(verifyApplicationPackage(f.root, { digest: f.digest }), /regular|symbolic/i);
});

test('rejects profile, runtime, traversal and case-collision mismatches', async t => {
  const f = await fixture(t);
  await assert.rejects(verifyApplicationPackage(f.root, { digest: f.digest, profile: { ...f.manifest.profile, audience: 'public' } }), /profile/i);
  f.manifest.nodeMajor++;
  await assert.rejects(verifyApplicationPackage(f.root, { digest: await f.save() }), /runtime/i);
  f.manifest.nodeMajor--;
  for (const path of ['private/../outside', 'private/back\\slash', 'private/renderer.mjs', 'PRIVATE/renderer.mjs']) {
    f.manifest.files.push({ path, hash: hash('fixture'), size: 7 });
    await assert.rejects(verifyApplicationPackage(f.root, { digest: await f.save() }), /path|collision/i);
    f.manifest.files.pop();
  }
  assert.equal((await readFile(join(f.root, 'private/renderer.mjs'), 'utf8')), 'fixture');
});
